### PR TITLE
#231 use different browser instances for before after methods

### PR DESCRIPTION
**Pull request by oomelianchuk**

### CODE REVIEW

**Review by wurzelkuchen — CHANGES_REQUESTED** [review]
Small formal things

File: src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/DontStartNewBrowserForCleanUp.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+@Retention(RUNTIME)
+@Target(
+{
+  TYPE, METHOD
+})
+public @interface DontStartNewBrowserForCleanUp
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/browser/multibrowser/DontStartNewBrowserForCleanUp.java]: javadock added

File: src/test/java/com/xceptance/neodymium/testclasses/softassertion/UseSoftAssertions.java [hunk 2]
```diff
@@ -39,4 +40,10 @@ public void validateSoftAssertion()
             $("#notFound4").click();
         });
     }
+
+    @After
+    public void after()
+    {
+        System.out.println("after");
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/softassertion/UseSoftAssertions.java]: removed

File: src/main/java/com/xceptance/neodymium/common/browser/BrowserAfterRunner.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            startNewBrowserForCleanUp = true;
+        }
+
+        if (!methodStartNewBrowserForCleanUp.isEmpty())
+        {
+            startNewBrowserForCleanUp = true;
+        }
+
+        // if @After method is annotated with @SuppressBrowser annotation, no new browser should be started
```
Comment (wurzelkuchen) [inline src/main/java/com/xceptance/neodymium/common/browser/BrowserAfterRunner.java:113]: the code does not reflect what the comment demands here or did I see this wrong?

File: src/main/java/com/xceptance/neodymium/common/browser/BrowserBeforeRunner.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        {
+            startNewBrowserForSetUp = true;
+        }
+
+        if (!methodStartNewBrowserForSetUp.isEmpty())
+        {
+            startNewBrowserForSetUp = true;
+        }
+
+        // if @Before method is annotated with @SuppressBrowser annotation, no new browser should be started
```
Comment (wurzelkuchen) [inline src/main/java/com/xceptance/neodymium/common/browser/BrowserBeforeRunner.java:121]: again the code does not do what the comment demands, does it? there is no check for suppressBrowser here

**Review by wurzelkuchen — CHANGES_REQUESTED** [review]
some additional notes

**Review by wurzelkuchen — APPROVED** [review]

**Pull request merged**